Lf 4260 patch and delete for soil amendment task

Duncan-Brain opened the pull request:
**Description**

This modifies the controllers to accept the new architecture changes in LF-4216 for Patch and Delete. (A follow up PR ensures the product endpoint is also compatible)

Changes:
- Complete (Update) endpoint allows deletions and insertions on `soil_amendment_task`
- Makes new checkTask middlewares for abandon, complete, and delete 
  - prior to this it was possible to complete or abandon tasks successfully without abandon information like date or duration
  - user_id was not required leading to faulty checks or lack of notifications.
  - You can see I had to call the other middleware in the checkComplete() due to the fact that in complete task it is optional to provide task details, but in other endpoints it is mandatory. There is probably other ways to do this but I thought this one was easiest for now.
- Tests:
  - fixes some test errors, `Other` type must exist in the db for the controller check on `other_purpose` notes. This is needed because of the super fun way we delete enum tables in the testEnvironment. 
  -  faker data should be greater than 0 for some fields (see updates below)
  - abandon_date required now
  - adds test for deleted or updated soil amendment task data

Updates not done (migrations needed but will add to migrations on LF-4246 to reduce # of migrations)
- greater than zero checks on the db level (Shubleen founds some negative values)
- was able to create multiple `soil_amendment_task_products` with same product_id and task_id this should be constrained

Reminders I saw about code smell:
- using `req.headers.user_id` vs `req.auth.user_id`
- middleware with `knex` calls are non atomic which isn't ideal, it would be nice to figure that out
- deleting enum tables from `test_farm` seems unnecessary, same with testing insertion on them

Jira link: [LF-4260](https://lite-farm.atlassian.net/browse/LF-4260)

**Type of change**

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Passes test case
- [ ] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files


[LF-4260]: https://lite-farm.atlassian.net/browse/LF-4260?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

SayakaOno reviewed (commented):
I'll submit what I've reviewed so far. It looks complicated, and I need some more time to understand. Or it would be easier if you could explain on a call.
Is there a test for "should be able to delete a product"?

Duncan-Brain commented on packages/api/src/controllers/taskController.js:
```diff
@@ -72,18 +72,36 @@ async function updateTaskWithCompletedData(
-    );
-  return task;
+  if (typeOfTask === 'soil_amendment_task') {
+    // Allows the insertion of missing data if no id present
```
A task id will always be present but for example a updated task product -- it would not yet have an id.

Duncan-Brain commented:
@SayakaOno happy to explain anything! I did not add a delete product test -- I may have assumed there already was one. I can check and add to other PR 4260b. Or did you mean soil_amendment_product?

SayakaOno commented:
@Duncan-Brain I meant the latter! I re-reviewed the test and now I understand what you did, thank you!
I'm not sure if we want to send `deleted: true` from the frontend though. Let's discuss it later!

Duncan-Brain commented:
Test body for complete: 
```
{
    "happiness":1,
    "duration": 15,
    "complete_date": "2024-06-10",
    "soil_amendment_task": {
        "task_id": 422,
        "method_id": 1,
        "soil_amendment_task_products": [
            {
            "id": 22,
            "deleted": true,
            "product_id": 35,
            "weight":30,
            "weight_unit": "kg",
            "application_rate_weight_unit": "kg/ha",
            "percent_of_location_amended": 100,
            "total_area_amended": 10,
            "purpose_relationships": [
                { "task_products_id": 22, "purpose_id": 5, "other_purpose": "nerr"}
            ]
            },
            {
            "id": 23,
            "product_id": 31,
            "weight":37,
            "weight_unit": "kg",
            "application_rate_weight_unit": "kg/ha",
            "percent_of_location_amended": 100,
            "total_area_amended": 10,
            "purpose_relationships": [
                { "task_products_id": 23, "purpose_id": 1 }
            ],
            }
        ]
    },
    "task_type_id": 6,
    "due_date": "2024-06-10"
}
```

Duncan-Brain commented:
I have added cascading delete based on discussion yesterday.

@SayakaOno  as part of this, for the request body I think I did not do the correct model architecture. As just mentioned in our slack discussion `task` is parent to `soil_amendment_task_products` . It is made this way to not give base properties to `soil_amendment_task` . So I have moved it up in expected request body. I hope it is not too inconvenient.

I think this makes sense as task should be parent to cross modules like taskLocations, managementTasks, and now taskProducts. 

 Things for monday :
 - Sayakas comments addressed
 - comments + cleanup
 - adjust update behaviour so that missing elements delete instead of requiring providing deleted row with deleted:true

kathyavini commented on packages/api/src/middleware/validation/checkTask.js:
```diff
@@ -0,0 +1,168 @@
+  };
+}
+
+const checkUpdateMiddlewareMap = {
```
Love this file and doing these checks with middleware! ❤️ 

Just checking a few things regarding this map:

1. Didn't quite follow what you meant here:

> You can see I had to call the other middleware in the checkComplete() due to the fact that in complete task it is optional to provide task details, but in other endpoints it is mandatory.

But as written, isn't this equivalent to just calling the middleware in sequence on the route?

```js
router.patch(
  '/complete/soil_amendment_task/:task_id',
  modelMapping['soil_amendment_task'],
  hasFarmAccess({ params: 'task_id' }),
  checkScope(['edit:task']),
  checkCompleteTask(), // without using the map
  checkSoilAmendmentTaskProducts(),
  taskController.completeTask('soil_amendment_task'),
);
```

I don't mind the map (as a reminder for which task types don't have a body content check in place yet) but I don't understand the task details part of what you said.

2. Just curious when you were thinking this middleware would go on all the other task routes. Will it be a ticket/task for right after release?

Duncan-Brain replied:
So I think as it is written if the `soil_amendment_task: {..}` was not provided in `task:{..}`. Then `checkSoilAmendmentTask()` would fail every time it is not provided. If it were chained on the route. Because that runs the check every time.

Since you can select "No changes were made to this task" that is a possibility.

This line bypasses `checkSoilAmendmentTask()` and moves to the controller if not defined:
```
 if (taskType in req.body) {
        checkUpdateMiddlewareMap[taskType]()(req, res, next);
      } else {
        next();
      }
```

That line was present in the previous middleware that was deleted `createOrPatchProduct()`: https://github.com/LiteFarmOrg/LiteFarm/pull/3208/commits/9de705c304b924c45aaa4b54aad2ca2a85e9cd9c

kathyavini replied:
Ohhhhh, okay. That makes sense, thank you 🙏

Duncan-Brain replied:
Oh forgot to respond to 2. 

I have been wondering if we should create release epics on JIRA for future releases so if we see something that should be added we can make a ticket and evaluate it in there. 

I see this as part of tasks redesign personally. Or if we have the chance to do some tech debt and firm up our API.

kathyavini commented on packages/api/tests/task.test.js:
```diff
@@ -1862,6 +1878,318 @@ describe('Task tests', () => {
+            expect(deletedRelationship[0].purpose_id).not.toBe(
+              deletedPurposeRelationship.purpose_id,
+            );
+            console.log('patch', completed_soil_amendment_task_products_purpose_relationship);
```
Stray `console.log()`?

Duncan-Brain commented on packages/api/src/controllers/managementPlanController.js:
```diff
@@ -404,10 +404,16 @@ const managementPlanController = {
+            }),
+          );
+
+          // await TaskModel.query(trx)
```
comments and cleanup on a friday late push haha -- thanks

Duncan-Brain commented on packages/api/src/models/taskModel.js:
```diff
@@ -414,16 +424,38 @@ class TaskModel extends BaseModel {
     }
   }
+
+  static async deleteTaskAndTaskProduct(user, task_id, trx) {
```
Nice! yah that name was awful -- love the new name. I took most of the suggestions. Kept the toLowerCase() in just one const line.

SayakaOno reviewed (commented):
I'm not sure if it has been like this, but I will provide a few things I found that shouldn't happen.
- Tasks can be created without `${type}_task: {}`
- When completing a soil amendment task, the id provided in the URL is ignored, and  the task with the `task_id` provided in the request body is completed instead. (I didn't test other types of tasks)

I will review the rest once the code is finalized!

Duncan-Brain commented on packages/api/src/controllers/taskController.js:
```diff
@@ -762,11 +763,16 @@ const taskController = {
       const checkTaskStatus = await TaskModel.getTaskStatus(task_id);
-      if (checkTaskStatus.complete_date || checkTaskStatus.abandon_date) {
+      if (checkTaskStatus?.complete_date || checkTaskStatus?.abandon_date) {
         return res.status(400).send('Task has already been completed or abandoned');
```
I considered that, but `checkTaskStatus()` is needed for `sendNotification()` so I chose not to duplicate the get request. If you do not mind then I can add it to the middleware!

SayakaOno replied:
But didn't you move the same code for `abandonTask` that also calls `sendNotification()`?
Maybe I'm saying something weird 😂

Duncan-Brain replied:
Probably .. haha I was of two minds about it. Consider it moved!

Duncan-Brain commented:
@SayakaOno 

I can try to close those loopholes for the new soil amendment but I think they are pre-existing. First one to allow for custom task probably.

Here is where task_id is overwritten by data: 
![Screenshot 2024-07-08 at 4 29 47 PM](https://github.com/LiteFarmOrg/LiteFarm/assets/30075973/9579ae71-91e4-4904-82d3-3c40c591c707)

Duncan-Brain commented:
Okay so I think for the part where:

> adjust update behaviour so that missing array elements delete instead of requiring providing deleted row with deleted:true

Either moving up the child to the correct parent fixed that or @kathyavini was right and it should work that way. Which was also my understanding but it wasnt working right when I made it. In either case it should be functioning now as expected:
- Deletion of product (by omission from body) -  soft deletes product and leaves purposes in tact
- Deletion of purposes (by omission from body) - hard deletes purpose

HOWEVER, I used a modifier on the model to return the right request body on `/complete` and it changed the `/delete` response to no longer include the deleted products. I think this is okay since we should maybe only be returning an empty "success" response as good practice from the delete endpoint. But how use base object property data like deleted is not consistent so I don't know if I did something wrong. Any opinions?

I closed those loopholes @SayakaOno mentioned but I think they were pre-existing.

SayakaOno commented:
Thank you Duncan! 🙌 
Could you please update the tests _"should be able to update a soil amendment task"_ and _"should not be able to delete the last product"_ that have `deleted = true`? 🙏

Duncan-Brain commented:
Thanks @SayakaOno forgot about that test!

SayakaOno reviewed (commented):
When completing a task, if I modify purposes, the existing `soil_amendment_task_product` gets `deleted: true` and I get a new record. Was that our decision?

Duncan-Brain commented on packages/api/tests/task.test.js:
```diff
@@ -1273,16 +1277,20 @@ describe('Task tests', () => {
           const data = {
             ...mocks.fakeTask({
-              [type]: { ...(await fakeTaskData[type](farm_id)) },
+              [type]: { ...fakeTaskData[type]() },
```
Yes thanks, leftover from when async was in taskData 🙏

Duncan-Brain commented on packages/api/tests/task.test.js:
```diff
@@ -1862,6 +1879,315 @@ describe('Task tests', () => {
+          async (err, res) => {
+            expect(res.status).toBe(400);
+
+            // Two active and one deleted task product should be present
```
Yeah it can be only one -- I was copying and pasting the other test to be fast with tests, but extra stuff is confusing and slow -- will remove extra.

Duncan-Brain commented on packages/api/tests/task.test.js:
```diff
@@ -1862,6 +1879,315 @@ describe('Task tests', () => {
+        // Delete abandonment reason to prevent validation error
+        delete createdTask.abandonment_reason;
+
+        // Remove all purpose relationships
+        const taskProductForDeletedPurpose = createdTaskProducts.find(
+          (product) => product.product_id == soilAmendmentProductTwo,
+        );
+        const deletedPurposeRelationship = taskProductForDeletedPurpose.purpose_relationships.pop();
+
+        taskProductForDeletedPurpose.purpose_relationships.push(deletedPurposeRelationship);
```
will remove extra

Duncan-Brain commented:
@SayakaOno 

It is expected that ...task_products without an id field will be deleted. And a new record for ...task_products without an id will be created. (As discussed in tech-daily it needs to be migrated to prevent multiple combinations of (product_id, task_id) to in the future be unique(product_id, task_id) but since there is not edit it is not urgent)

If the ...task_products id is provided it should be updated correctly. Is that what you see?

SayakaOno commented:
@Duncan-Brain Ah sorry, I was talking without seeing your sample request 🙏 It worked as expected!
> soil_amendment_task_products without the id field will be created as new

> in the future be unique(product_id, task_id)

Should we assume that a work around is required on the frontend in this scenario then?

1. Create a task with product A and B.
2. When completing, remove product A once and add product A again.

On the backend, the existing record will be marked as "deleted", so a new record for product A will not be created due to a database error.
We will need to send `soil_amendment_task_products` `id` with the new record for product A, correct? It seems complicated...

Duncan-Brain commented:
@SayakaOno I think in that scenario a new product is created successfully (but it is a duplicate). Do you see a db error?

What would need to happen on the frontend is : 
- Product A (id: X) is removed
- Product A (id: to be determined) is re-added
- Send to db is successful ==> delete product with id:X, add returned id for new product A.

SayakaOno commented:
It works "for now", right?
Once "unique(product_id, task_id)" is implemented, I thought duplicates would not be allowed.

Duncan-Brain commented:
Yes, I still hope I can squeeze in the migrations before release but it sounded like it was not important... 

My intention is to migrate and keep the `deleted: true` or most recently `created_at` row deleting the others. Then fix the controller logic to check if it exists and update accordingly at that time. 

I could do the controller logic part earlier... but would have to ask you and Joyce to delete records by rolling back recent migrations so it works as expected.

SayakaOno commented:
Just to clarify, "fix the controller logic to check if it exists and update accordingly at that time" - this will enable the scenario below to work, correct?

> 1. Create a task with product A and B.
> 2. When completing, remove product A once and add product A again.

I don't think we need the migration, but the controller logic is worth including in this release to avoid writing a migration when adding the unique constraint!

SayakaOno reviewed (commented):
Looks good! 🎉

Duncan-Brain commented:
@SayakaOno Okay I can add in now, are you and @kathyavini ok to rollback LF-4216 migrations to destroy your junk data before testing again? (you could also manually delete duplicates, then purpose relationships will have also need manual deletions, and making sure true if only one task product record exists, and not deleting all purposes -- if some are missed then there might be unexpected bugs but should be fine if done that way too) .

kathyavini reviewed (approved):
I didn't fully the adding-removing edge case most recently discussed so I'm not sure what is coming in the forthcoming controller update, but for me this works exactly as I was expecting -- and it has also improved the general task controller flow so much!! ❤️
